File: Python/ByTier/D/2002_D1_DFS_Checker_Easy_Version.py
# Problem: CF 2002 D1 - DFS Checker (Easy Version)
# https://codeforces.com/contest/2002/problem/D1

"""
D1. DFS Checker (Easy Version)

Purpose:
This code determines whether a given permutation of vertices in a perfect binary tree represents a valid Depth-First Search (DFS) order. For each query, it swaps two elements in the permutation and checks if the resulting sequence can be a DFS traversal.

The algorithm uses a special structure for a perfect binary tree to efficiently compute parent-child relationships and then verifies conditions based on the idea that in a valid DFS order, each node's value must be at most half of its parent's value (under specific indexing), which ensures that a DFS traversal is possible from that node upwards and downward.

Algorithms/Techniques:
- Perfect Binary Tree Construction using bit-lengths and index mapping
- Preprocessing to check "validity" conditions for DFS
- Efficient query handling with partial updates

Time Complexity: O(n + q * sqrt(n))
Space Complexity: O(n)

Note:
The tree is represented implicitly via its parent relationships.
Each node's children are determined using binary indexing (child1 = i+1, child2 = i + 2^(k - level[i]))
Valid DFS order condition: For each vertex i (except root), p[i] // 2 == p[parent[i]]
"""

import sys

input = lambda: sys.stdin.readline().rstrip()


def solve_case():
    n, q = map(int, input().split())
    _ = list(map(int, input().split()))
    p = list(map(int, input().split()))
    k = n.bit_length()  # Determine height of the perfect binary tree
    level = [-1] * n
    parent = [-1] * n
    child1 = [-1] * n
    child2 = [-1] * n
    level[0] = 1
    parent[0] = -1
    for i in range(n):
        if level[i] < k:
            child1[i] = i + 1  # First child
            child2[i] = i + 2 ** (k - level[i])  # Second child
            level[child1[i]] = level[child2[i]] = level[i] + 1
            parent[child1[i]] = parent[child2[i]] = i
    
    # ok[i] represents if the property holds for node i
    ok = [0] * n
    for i in range(n):
        if parent[i] != -1:
            ok[i] = int(p[i] // 2 == p[parent[i]])  # Property for valid DFS order

    sum_ok = sum(ok)  # Sum of all valid nodes
    ans = ["NO"] * q
    for i in range(q):
        x, y = map(lambda x: int(x) - 1, input().split())  # Convert to 0-indexed
        
        s = set()  # Set to store affected nodes
        
        if parent[x] != -1:
            s |= {x}  # Add node x to check
        if child1[x] != -1:
            s |= {child1[x], child2[x]}  # Add children of x to check
        if parent[y] != -1:
            s |= {y}  # Add node y to check
        if child1[y] != -1:
            s |= {child1[y], child2[y]}  # Add children of y to check
            
        # Update sum_ok by removing old values
        for j in s:
            sum_ok -= ok[j]
            ok[j] = 0

        # Simulate swap in permutation
        p[x], p[y] = p[y], p[x]

        # Update ok array and sum_ok based on the new values
        for j in s:
            ok[j] = int(p[j] // 2 == p[parent[j]])  # Update validity condition
            sum_ok += ok[j]
            
        if sum_ok == n - 1:  # N-1 valid nodes (except root)
            ans[i] = "YES"
    return "\n".join(ans)


ans = []
for _ in range(int(input())):
    ans += [str(solve_case())]
print("\n".join(ans))


# https://github.com/VaHiX/codeForces/

File: CPP/ByTier/D/2002_D1_DFS_Checker_Easy_Version.cpp
// Problem: CF 2002 D1 - DFS Checker (Easy Version)
// https://codeforces.com/contest/2002/problem/D1

/*
 * Purpose: Determines if a given permutation of vertices of a perfect binary tree
 *          represents a valid DFS traversal order after a series of swaps.
 *
 * Algorithm:
 *   - Build a perfect binary tree with n nodes (n = 2^k - 1)
 *   - For each query:
 *     - Swap two elements in the permutation
 *     - Check if the resulting permutation is a valid DFS order
 *   - Uses a bottom-up approach with recursive DFS check on the tree
 *   - Maintains a global "ok" array to track valid DFS orders for each subtree
 *   - Uses a "rev" array to map vertex values to their positions in the permutation
 *   - Uses "sz" array to store size of each subtree to determine valid child order
 *
 * Time Complexity: O(n log n + q * log n)
 *   - O(n) to initialize tree information
 *   - O(log n) per query for DFS validation (upward propagation)
 * 
 * Space Complexity: O(n)
 *   - Stores arrays for tree structure and validation info
 */
#include <algorithm>
#include <iostream>
#include <vector>
#include <utility>

using namespace std;
using ll = long long;
using ull = unsigned long long;
using uint = unsigned int;
using db = double;
int n;
vector<bool> ok;        // Tracks if subtree rooted at i is valid DFS order
vector<int> sz, rev, p;  // sz: subtree sizes, rev: reverse mapping of permutation, p: current permutation

// Initialize the valid DFS status for each node
void initdfs(int v) {
  if (v * 2 >= n) {     // Base case: leaf node
    ok[v] = true;
    return;
  }
  initdfs(2 * v);       // Recursively process left child
  initdfs(2 * v + 1);   // Recursively process right child
  
  // If either child is invalid, then current node is also invalid
  if (!ok[2 * v] || !ok[2 * v + 1]) {
    ok[v] = false;
    return;
  }
  
  // Check if the current node follows valid DFS order based on children
  // Case 1: Left child comes first
  if ((rev[v] + 1 == rev[2 * v]) &&
      (rev[v] + 1 + sz[2 * v] == rev[2 * v + 1])) {
    ok[v] = true;
    return;
  }
  // Case 2: Right child comes first
  if ((rev[v] + 1 == rev[2 * v + 1]) &&
      (rev[v] + 1 + sz[2 * v + 1] == rev[2 * v])) {
    ok[v] = true;
    return;
  }
  ok[v] = false;        // Neither case holds, invalid DFS
}

// Update the DFS validity for all ancestral nodes after a swap
void upd(int a, int b) {
  if (a < b) {
    swap(a, b);         // Ensure a >= b
  }
  if (a == b) {
    b /= 2;             // Handle case where a == b (special handling)
  }
  if (a == 0) {
    return;             // Base case reached
  }
  if (a * 2 >= n) {
    upd(a / 2, b);      // If a is leaf, update parent
    return;
  }
  
  // If either child is invalid, current node is also invalid
  if (!ok[2 * a] || !ok[2 * a + 1]) {
    ok[a] = false;
    upd(a / 2, b);
    return;
  }
  
  // Check if DFS order is valid based on children placement
  if ((rev[a] + 1 == rev[2 * a]) &&
      (rev[a] + 1 + sz[2 * a] == rev[2 * a + 1])) {
    ok[a] = true;
    upd(a / 2, b);
    return;
  }
  if ((rev[a] + 1 == rev[2 * a + 1]) &&
      (rev[a] + 1 + sz[2 * a + 1] == rev[2 * a])) {
    ok[a] = true;
    upd(a / 2, b);
    return;
  }
  ok[a] = false;
  upd(a / 2, b);
}

// Process each test case
void sol() {
  int q, a, b;
  cin >> n >> q;
  sz.resize(n + 1), ok.resize(n + 1), rev.resize(n + 1), p.resize(n + 1);
  
  int cur = n, st = 2;  // st tracks boundary of current level in the tree
  for (int i = 1; i <= n; i++) {
    if (i == st) {
      cur /= 2;         // When we reach the end of a level, decrease level size
      st *= 2;
    }
    sz[i] = cur;        // Fill subtree sizes
  }
  
  // Read parent information (not used in code as it's confirmed to be floor(i/2))
  for (int i = 0; i < n - 1; i++) {
    cin >> a;
  }
  
  // Initialize permutation and reverse mapping
  for (int i = 1; i <= n; i++) {
    cin >> p[i];
    rev[p[i]] = i;      // Map vertex value to its index in permutation
  }
  
  initdfs(1);           // Start initialization from root
  
  // Process queries
  for (int i = 0; i < q; i++) {
    cin >> a >> b;
    swap(p[a], p[b]);   // Swap values in permutation
    rev[p[a]] = a;      // Update reverse mapping
    rev[p[b]] = b;
    upd(p[a], p[b]);    // Update validity of affected nodes
    if (ok[1]) {        // Print result for root
      cout << "YES\n";
    } else {
      cout << "NO\n";
    }
  }
}

int main() {
  ios_base::sync_with_stdio(false);
  cin.tie(nullptr);
  int t = 1;
  cin >> t;
  while (t--) {
    sol();
  }
}


// https://github.com/VaHiX/CodeForces/